pkg/core/notary_test.go

```diff
@@ -0,0 +1,668 @@
+			_, err := bc.genBlocks(int(nvbDiffFallback))
+			require.NoError(t, err)
+			unlucky := requests[nSigs-1]
+			ntr1.OnFallbackRemoval(unlucky)
+			// then m out of n fallbacks should be completed
+			finalizeWithError = false
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkMultisigTx(t, nSigs, requests, len(requests), false)
+			checkFallbackTxs(t, requests[:nSigs-1], true)
+			require.Nil(t, completedTxes[unlucky.FallbackTransaction.Hash()])
+			// but the rest (n-m) out of n fallbacks shouldn't be completed because main tx has been collected by the moment they were sent
+			checkFallbackTxs(t, requests[nSigs:], false)
+
+			// make another request
+			finalizeWithError = true
+			requests = checkCompleteMultisigRequest(t, nSigs, nKeys, false)
+
+			// make fallbacks valid and then remove all of them
+			_, err = bc.genBlocks(int(nvbDiffFallback))
+			require.NoError(t, err)
+			for i := range requests {
+				ntr1.OnFallbackRemoval(requests[i])
+			}
+			// then the whole request should be removed, i.e. there are no completed transactions
+			finalizeWithError = false
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkMultisigTx(t, nSigs, requests, len(requests), false)
+			checkFallbackTxs(t, requests, false)
+
+			// it's OK to remove unexisting fallback transaction, i.e. there still shouldn't be any completed transactions after this
+			ntr1.OnFallbackRemoval(requests[0])
+			require.NoError(t, bc.AddBlock(bc.newBlock()))
+			checkMultisigTx(t, nSigs, requests, len(requests), false)
+			checkFallbackTxs(t, requests, false)
+		})
+	})
+}
+
+func TestNotary_UpdateNotaryNodes(t *testing.T) {
+	bc := newTestChain(t)
+	defer bc.Close()
+	acc1, ntr1 := getTestNotary(t, bc, "./testdata/notary1.json", "one")
+	randomKey, err := keys.NewPrivateKey()
+	require.NoError(t, err)
+	// set account for the first time
+	ntr1.UpdateNotaryNodes(keys.PublicKeys{acc1.PrivateKey().PublicKey()})
+
+	t.Run("account is already set", func(t *testing.T) {
+		ntr1.UpdateNotaryNodes(keys.PublicKeys{acc1.PrivateKey().PublicKey(), randomKey.PublicKey()})
+	})
+
+	t.Run("another account from the same wallet", func(t *testing.T) {
+		t.Run("good config password", func(t *testing.T) {
+			w, err := wallet.NewWalletFromFile(path.Join(notaryModulePath, "./testdata/notary1.json"))
+			require.NoError(t, err)
+			require.NoError(t, w.Accounts[1].Decrypt("one"))
+			ntr1.UpdateNotaryNodes(keys.PublicKeys{w.Accounts[1].PrivateKey().PublicKey()})
+		})
+		t.Run("bad config password", func(t *testing.T) {
+			w, err := wallet.NewWalletFromFile(path.Join(notaryModulePath, "./testdata/notary1.json"))
+			require.NoError(t, err)
+			require.NoError(t, w.Accounts[2].Decrypt("four"))
+			ntr1.UpdateNotaryNodes(keys.PublicKeys{w.Accounts[2].PrivateKey().PublicKey()})
+			// TODO: check that account is nil (after moving test to services pkg)
+		})
+	})
+
+	t.Run("unknown account", func(t *testing.T) {
+		ntr1.UpdateNotaryNodes(keys.PublicKeys{randomKey.PublicKey()})
```

Comment:
Same here. `NotPanics`? Or check that notary module is suspend?

Reply:
I'll move this test and add checks to it. It's not completed yet.